spatialos-sdk/src/worker/schema.rs

```diff
@@ -0,0 +1,606 @@
+    Schema_AddFixed64,
+    Schema_AddFixed64List,
+    Schema_GetFixed64List,
+);
+generate_schema_primitive!(
+    i32,
+    SchemaSfixed32,
+    Schema_GetSfixed32,
+    Schema_IndexSfixed32,
+    Schema_GetSfixed32Count,
+    Schema_AddSfixed32,
+    Schema_AddSfixed32List,
+    Schema_GetSfixed32List,
+);
+generate_schema_primitive!(
+    i64,
+    SchemaSfixed64,
+    Schema_GetSfixed64,
+    Schema_IndexSfixed64,
+    Schema_GetSfixed64Count,
+    Schema_AddSfixed64,
+    Schema_AddSfixed64List,
+    Schema_GetSfixed64List,
+);
+generate_schema_primitive!(
+    u32,
+    SchemaEnum,
+    Schema_GetEnum,
+    Schema_IndexEnum,
+    Schema_GetEnumCount,
+    Schema_AddEnum,
+    Schema_AddEnumList,
+    Schema_GetEnumList,
+);
+
+// NOTE: It's safe to treat the `EntityId` array as an `i64` array because
+// `EntityId` is marked `repr(transparent)`, and so is identical to an `i64` for
+// the purpose of FFI.
+impl_primitive_field!(
+    EntityId,
+    EntityId,
+    Schema_GetEntityId,
+    Schema_IndexEntityId,
+    Schema_GetEntityIdCount,
+    Schema_AddEntityId,
+    Schema_AddEntityIdList,
+    Schema_GetEntityIdList,
+);
+
+// NOTE: Bools in a schema object are guaranteed to be represented as either 0 or 1,
+// so it's safe to directly convert the returned values to Rust a `bool` without
+// intermediate conversions.
+impl_primitive_field!(
+    bool,
+    bool,
+    Schema_GetBool,
+    Schema_IndexBool,
+    Schema_GetBoolCount,
+    Schema_AddBool,
+    Schema_AddBoolList,
+    Schema_GetBoolList,
+);
+
+impl<T: IndexedField> SchemaFieldSealed for Option<T> {}
+
+impl<T: IndexedField> SchemaField for Option<T> {
+    type RustType = Option<T::RustType>;
+
+    fn add_field(object: &mut Object, field: FieldId, value: &Self::RustType) {
+        if let Some(value) = value {
+            T::add_field(object, field, value);
+        }
+    }
+
+    fn get_field(object: &Object, field: FieldId) -> Self::RustType {
+        match T::field_count(object, field) {
+            0 => None,
+            _ => Some(T::get_field(object, field)),
+        }
+    }
+
+    fn has_update(update: &ComponentUpdate, field: FieldId) -> bool {
+        T::field_count(update.fields(), field) > 0 || update.field_cleared(field)
+    }
+
+    fn get_update(update: &ComponentUpdate, field: FieldId) -> Option<Self::RustType> {
+        if update.field_cleared(field) {
+            Some(None)
+        } else {
+            Self::get_field(update.fields(), field).map(Some)
+        }
+    }
+
+    fn add_update(update: &mut ComponentUpdate, field: FieldId, value: &Self::RustType) {
+        match value {
+            Some(value) => {
+                update.add_field::<T>(field, value);
+            }
+
+            None => update.add_cleared(field),
+        }
+    }
+}
+
+impl<K, V> SchemaFieldSealed for BTreeMap<K, V>
+where
+    K: IndexedField,
+    V: IndexedField,
+    K::RustType: Ord,
+{
+}
+
+impl<K, V> SchemaField for BTreeMap<K, V>
```

Comment:
Instead of implementing it on an instantiation of `BTreeMap` with the marker types as keys and values, can we introduce something like `pub struct SchemaMap<K, V>;` as a kind of "marker map"?

I guess this would change the usage to something like:
```
data.fields().field::<SchemaMap<String, i32>>(2);
```
instead of
```
data.fields().field::<BTreeMap<String, i32>>(2);
```